<a href="https://colab.research.google.com/github/ljs7463/study-repository/blob/master/%ED%8C%8C%EC%9D%B4%EC%8D%AC%EC%97%B0%EC%8A%B5(beginner)/Data%20analysis%20practice/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
series : 1차원으로 이루어진 데이터 배열
DataFrame : 2차원으로 이루어진 데이터 배열

In [ ]:
import pandas as pd
pd

<module 'pandas' from '/usr/local/lib/python3.6/dist-packages/pandas/__init__.py'>

# Pandas 의 Series 와 DataFrame

1차원, 1개의 column은 Series라고 합니다.

In [ ]:
pd.Series([1, 2, 3, 4])

0    1
1    2
2    3
3    4
dtype: int64

In [ ]:
a = [1, 2, 3, 4]
pd.Series(a)

0    1
1    2
2    3
3    4
dtype: int64

In [ ]:
mylist = [1, 2, 3, 4]
pd.Series(mylist)

0    1
1    2
2    3
3    4
dtype: int64

# DataFrame

방법 1. list로 만들기

In [ ]:
company1 = [['삼성', 2000, '스마트폰'],
           ['현대', 1000, '자동차'],
           ['네이버', 500, '포털']]

In [ ]:
pd.DataFrame(company1)


0     1     2
0   삼성  2000  스마트폰
1   현대  1000   자동차
2  네이버   500    포털

In [ ]:
df1 = pd.DataFrame(company1)
df1

0     1     2
0   삼성  2000  스마트폰
1   현대  1000   자동차
2  네이버   500    포털

제목 컬럼 만들기

In [ ]:
df1.columns = ['기업명','매출액','업종']
df1.index = ['첫 번째', '두 번째', '세 번째']
df1 

기업명   매출액    업종
첫 번째   삼성  2000  스마트폰
두 번째   현대  1000   자동차
세 번째  네이버   500    포털

방법2. dict로 만들기

In [ ]:
company2 = {'기업명':['삼성','현대','네이버'],
           '매출액': [2000, 1000, 500],
           '업종': ['스마트폰', '자동차', '포털']
           }
df2 = pd.DataFrame(company2)
df2

기업명   매출액    업종
0   삼성  2000  스마트폰
1   현대  1000   자동차
2  네이버   500    포털

index를 특정 column으로 지정하기

In [ ]:
df1
df1.index = df1['기업명']
df1

기업명   매출액    업종
기업명                 
삼성    삼성  2000  스마트폰
현대    현대  1000   자동차
네이버  네이버   500    포털

series = column 이라고 하였습니다. 확인해 볼까요?

In [ ]:
df1['매출액']

In [ ]:
type(df1['매출액'])

# CSV 파일 읽어오기

CSV 파일 이란?
   - Comma Separated Value의 약어로써, 쉼표로 구분된 파일(tab으로도 가능)
   - 엑셀을 로딩할 수도 있지만, 쉼표로 구분된 파일이 훨씬 가볍다
   - 공공데이터포털 에서도 CSV 포맷의 파일을 제공한다.

1-1. 파일1) 로컬에서 파일 읽어오기(복습) - 초심자가 진행하기에 번거롭기 때문에 비추천

In [ ]:
from google.colab import files
myfile = files.upload()

KeyboardInterrupt: ignored

In [ ]:
import io

In [ ]:
pd.read_csv(io.BytesIO(myfile['korean-idol.csv'])) 

1-2. 방법2) 구글 드라이브에 있는 샘플 파일 읽어오기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
filename = '/content/drive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.csv'
pd.read_csv(filename)  # pd.read_csv() 판다스 에서 csv파일 읽어 오기

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

1-3. 방법3) 파일 URL로 부터 바로 읽어오기

In [ ]:
pd.read_csv('http://bit.ly/ds-korean-idol')

1. 엑셀 파일 읽어오기


엑셀 파일을 읽어 올 때는 read_excel()함수를 사용합니다.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
import pandas as pd
filename = '/content/drive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.xlsx'

In [ ]:
pd.read_excel(filename)

1. 파일 읽어오기(csv)

In [ ]:
import pandas as pd
df =pd.read_csv('/content/drive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.csv')
df

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

2. 기본 정보 알아보기 (index, column, info)

2-1. column(열) 출력하기

In [ ]:
df.columns

Index(['이름', '그룹', '소속사', '성별', '생년월일', '키', '혈액형', '브랜드평판지수'], dtype='object')

2-2. column(열) 이름 재정의하기

In [ ]:
new_col = ['name', '그룹', '소속사', '성별', '생년월일', '키', '혈액형', '브랜드평판지수'] # 기존 cloumn과 같은 수
df.columns = new_col
df.columns

Index(['이름', '그룹', '소속사', '성별', '생년월일', '키', '혈액형', '브랜드평판지수'], dtype='object')

In [ ]:
df

name     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

2-2. index(행) 출력하기

In [ ]:
df.index

RangeIndex(start=0, stop=15, step=1)

2-3. info는 기본적인 row(행)의 정보와 데이터 타입을 알려줍니다.

tip. info 메소드는 주로 빠진 값(null값)과 데이터 타입을 볼 때 활용합니다.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   이름       15 non-null     object 
 1   그룹       14 non-null     object 
 2   소속사      15 non-null     object 
 3   성별       15 non-null     object 
 4   생년월일     15 non-null     object 
 5   키        13 non-null     float64
 6   혈액형      15 non-null     object 
 7   브랜드평판지수  15 non-null     int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 1.1+ KB


3. 통계 정보 알아보기(describe)

In [ ]:
df.describe() #산술 연산만 할 수 있는 column 만 출력한다.

키       브랜드평판지수
count   13.000000  1.500000e+01
mean   175.792308  5.655856e+06
std      5.820576  2.539068e+06
min    162.100000  2.925442e+06
25%    174.000000  3.712344e+06
50%    177.000000  4.668615e+06
75%    179.200000  7.862214e+06
max    183.000000  1.052326e+07

4. 형태(shape) 알아보기
  - shape는 tuple 형태로 반환되며, 첫번째는 row, 두번째는 column의 숫자를 의미합니다.

In [ ]:
df.shape

(15, 8)

5. 상위 5개, 하위 5개의 정보만 보기

가장 자주쓰는 메소드 중 하나인 head(), tail()입니다.
  - head() 상위 5개 row출력
  - tail() 하위 5개 row출력
  - head(3) 상위 3개 row출력
  - tail(2) 하위 2개 row출력

In [ ]:
df.head()

In [ ]:
df.tail()

In [ ]:
df.head(3)

7. 정렬하기

7-1. 오름차순 정렬(default)

In [ ]:
df.sort_index()

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

7-2.  내림차순 index 정렬

In [ ]:
df.sort_index(ascending=False)

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260

8. column 별로 정렬 

8-1. 오름차순 정렬

In [ ]:
df.sort_values(by = '키')

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661

8-2. 내림차순 정렬

In [ ]:
df.sort_values(by = '키',ascending=False )

[BONUS] 복수정렬

In [ ]:
df.sort_values(by = ['키', '브랜드평판지수']) #키순서대로 하고 같은 거는 브랜드평판지수순서로

In [ ]:
df.sort_values(by = ['키', '브랜드평판지수'], ascending=False)

1. Column을 선택하는 방법

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
df= pd.read_csv('/content/drive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.csv')
df.head()

In [ ]:
df['혈액형'] # 추천

0      A
1      A
2      A
3     AB
4      A
5      A
6      O
7      B
8      O
9      A
10     A
11     B
12    AB
13     O
14     O
Name: 혈액형, dtype: object

In [ ]:
df.혈액형

2. 범위 선택(range selection)


2-1. 단순 index에 대한 범위 선택

In [ ]:
df[:3] # = df.head(3)

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745

2-2 . loc

(앞으로 굉장히 자주 사용합니다. 반드시 손에 익숙해 지도록 합시다!)

In [ ]:
df.loc[:,'이름'] # row 전체 컬럼 은 이름만

0       지민
1     지드래곤
2     강다니엘
3        뷔
4       화사
5       정국
6       민현
7       소연
8        진
9      하성운
10      태연
11     차은우
12      백호
13      JR
14      슈가
Name: 이름, dtype: object

In [ ]:
df.loc[:, ['이름', '생년월일']]

이름        생년월일
0     지민  1995-10-13
1   지드래곤  1988-08-18
2   강다니엘  1996-12-10
3      뷔  1995-12-30
4     화사  1995-07-23
5     정국  1997-09-01
6     민현  1995-08-09
7     소연  1998-08-26
8      진  1992-12-04
9    하성운  1994-03-22
10    태연  1989-03-09
11   차은우  1997-03-30
12    백호  1995-07-21
13    JR  1995-06-08
14    슈가  1993-03-09

In [ ]:
df.loc[3:8, ['이름','생년월일']] 
# pandas에서 loc는 a:b 에서 b이하로 가져옴 즉 b포함해서 가져온다. numpy 는 a: b 일경우 b미만

이름        생년월일
3   뷔  1995-12-30
4  화사  1995-07-23
5  정국  1997-09-01
6  민현  1995-08-09
7  소연  1998-08-26
8   진  1992-12-04

In [ ]:
df.loc[2:5, '이름':'생년월일']

이름     그룹  소속사  성별        생년월일
2  강다니엘    NaN  커넥트  남자  1996-12-10
3     뷔  방탄소년단  빅히트  남자  1995-12-30
4    화사    마마무  RBW  여자  1995-07-23
5    정국  방탄소년단  빅히트  남자  1997-09-01

2-3. iloc(position으로 색인)

(i는 인덱스 즉 번호로 가지고 오고 싶을 때 사용)

In [ ]:
df.iloc[:,[0,2]]

이름      소속사
0     지민      빅히트
1   지드래곤       YG
2   강다니엘      커넥트
3      뷔      빅히트
4     화사      RBW
5     정국      빅히트
6     민현     플레디스
7     소연       큐브
8      진      빅히트
9    하성운  스타크루이엔티
10    태연       SM
11   차은우     판타지오
12    백호     플레디스
13    JR     플레디스
14    슈가      빅히트

In [ ]:
df.iloc[1:5,[0,2]] 
#iloc에서는 다시 a  : b 에서 b미만이 된다

이름  소속사
1  지드래곤   YG
2  강다니엘  커넥트
3     뷔  빅히트
4    화사  RBW

In [ ]:
df.iloc[1:5,1:4] 

그룹  소속사  성별
1     빅뱅   YG  남자
2    NaN  커넥트  남자
3  방탄소년단  빅히트  남자
4    마마무  RBW  여자

3. Boolean Indexing - 조건을 활용한 색인

Boolean Indexing은 Numpy에서 배웠 Boolean 인덱싱과 같은 원리입니다.

In [ ]:
df['키'] > 180
# df.loc[df['키']>180]

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10    False
11     True
12    False
13    False
14    False
Name: 키, dtype: bool

위 처럼 Boolean을 이용해서 구분하는 것으로 끝나는게 아닌 원하는 즉 색출하고자 하는 데이터 를 뽑아내는것이 목적 

Boolean Index로 받은 index 를 활용해서 True인 값만 색인해 낼 수 있습니다.

  - df[] 꺾쇠로 감싸주고, 그 안에 Boolean Index를 넣어주시면 됩니다.

In [ ]:
df[df['키'] < 170]

이름   그룹      소속사  성별        생년월일      키 혈액형  브랜드평판지수
4   화사  마마무      RBW  여자  1995-07-23  162.1   A  7650928
9  하성운   핫샷  스타크루이엔티  남자  1994-03-22  167.1   A  4036489

위와 같은 방법이 매우 간편한 방법이기는 하지만, 모든 column을 출력해야만 한다는 한계가 있습니다.

  - 특정 column과 같이 색인해 내고 싶을땐 어떻게 할까요?

In [ ]:
#df[df['키'] > 180 , '이름'] 이렇게 하면 에러가 남 


해결방법 1. 맨 뒤에 출력할 column 붙히기

In [ ]:
df[df['키']>180]['이름']

In [ ]:
df[df['키'] > 180][['이름', '키']]

해결방법 2. loc를 활용 (추천)

In [ ]:
df.loc[df['키']>180, '이름']

6      민현
11    차은우
Name: 이름, dtype: object

In [ ]:
df.loc[df['키']>180, '이름' : '성별']

이름    그룹   소속사  성별
6    민현  뉴이스트  플레디스  남자
11  차은우  아스트로  판타지오  남자

In [ ]:
df.loc[df['키']>180, ['이름','키']] #따로 할 떄는 리스트로 

이름      키
6    민현  182.3
11  차은우  183.0

4. isin을 활용한 색인

isin을 활용한 색인은 내가 조건을 걸고자 하는 값이 내가 정의한 list에 있을 때만 색인하려는 경우에 사용합니다.

In [ ]:
my_condition = ['플레디스', 'SM']

In [ ]:
df['소속사'].isin(my_condition) #소속사라는 컬럼을 지정하고(지정!!) 그 안에 isin()안에 있는 값들이 있어?

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10     True
11    False
12     True
13     True
14    False
Name: 소속사, dtype: bool

In [ ]:
df.loc[df['소속사'].isin(my_condition),'소속사']

6     플레디스
10      SM
12    플레디스
13    플레디스
Name: 소속사, dtype: object

5. 결측값(Null)알아보기

5-1. NaN 값에 대하여

  - Null 값은 비어있는 값, 고급 언어로 결측값입니다.
  - pandas 에서는 NaN => Not a Number로 표기 된 것을 확인해 볼 수 있습니다.

In [ ]:
df # 결측값 즉, NaN값은 데이터다룸에 있어서 중요한 요소로 작용을 한다.

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

info()로 NaN값, 즉 빠진 데이터가 어디에 있는지 쉽게 요약정보로 확인할 수 있습니다. (빠진데이터를 처리하는것도 중요하고 처리하는 방식 또한 중요하다)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   이름       15 non-null     object 
 1   그룹       14 non-null     object 
 2   소속사      15 non-null     object 
 3   성별       15 non-null     object 
 4   생년월일     15 non-null     object 
 5   키        13 non-null     float64
 6   혈액형      15 non-null     object 
 7   브랜드평판지수  15 non-null     int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 1.1+ KB


5-2.결측 값 다루기

Boolean 인덱싱

In [ ]:
df.isna() #isna 빠진값이야? #하지만 이렇게 하면 보기가 여전히 불편하다

이름     그룹    소속사     성별   생년월일      키    혈액형  브랜드평판지수
0   False  False  False  False  False  False  False    False
1   False  False  False  False  False  False  False    False
2   False   True  False  False  False  False  False    False
3   False  False  False  False  False  False  False    False
4   False  False  False  False  False  False  False    False
5   False  False  False  False  False  False  False    False
6   False  False  False  False  False  False  False    False
7   False  False  False  False  False   True  False    False
8   False  False  False  False  False  False  False    False
9   False  False  False  False  False  False  False    False
10  False  False  False  False  False   True  False    False
11  False  False  False  False  False  False  False    False
12  False  False  False  False  False  False  False    False
13  False  False  False  False  False  False  False    False
14  False  False  False  False  False  False  False    False

info() 메소드를 통해서 전체적으로 어떤 column에 빠진 데이터가 있는지 알아 보았습니다. 또한, 다음과 같이 특정 column에서 빠진값을 색출해 낼 수 있습니다.

1. Boolean 인덱싱으로 True가 return되는 값이 NaN이라는 것을 알 수 있습니다.

In [ ]:
df['그룹'].isnull() #isnull로도 가능


0     False
1     False
2      True
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
Name: 그룹, dtype: bool

2. NaN값만 색출해내기

In [ ]:
df['그룹'][df['그룹'].isnull()]
# df.loc[df['그룹'].isnull(), '그룹']  # 이렇게도 가능

2    NaN
Name: 그룹, dtype: object

In [ ]:
df['그룹'][df['그룹'].isna()]

3. NaN이 아닌 값에 대하여 Boolean인덱싱(비어있지 않은값도 추출해야할 경우도 있음)

In [ ]:
df['그룹'].notnull()
# df['그룹'].notna()

0      True
1      True
2     False
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10     True
11     True
12     True
13     True
14     True
Name: 그룹, dtype: bool

4. NaN이 아닌 값만 색출해 내기

In [ ]:
df['그룹'][df['그룹'].notnull()]

0     방탄소년단
1        빅뱅
3     방탄소년단
4       마마무
5     방탄소년단
6      뉴이스트
7       아이들
8     방탄소년단
9        핫샷
10     소녀시대
11     아스트로
12     뉴이스트
13     뉴이스트
14    방탄소년단
Name: 그룹, dtype: object

In [ ]:
df.loc[df['그룹'].isnull()]
#df.loc[df['그룹'].isnull(),['키', '혈액형']]

이름   그룹  소속사  성별        생년월일      키 혈액형  브랜드평판지수
2  강다니엘  NaN  커넥트  남자  1996-12-10  180.0   A  8273745

1. copy (복사)

copy는 단어 그대로, dataframe을 복사할 때 사용합니다.

In [ ]:
df.head()

df를 새로운 변수 (new_df)에 대입해 주고, 새로운 변수에서 값을 바꾼다면, 원래의 DataFrame의 값은 어떻게 될까요?

In [ ]:
new_df = df.copy()
new_df['이름'] = 0
df.head

이렇게 되는 이유는 같은 메모리 주소를 참조하기 때문이다.
즉, 여러가지 실험을 해볼 때 쓰는용도 

굉장히 많은 분들이 실수를 하시는 부분입니다.

원본 데이터를 유지 시키고, 새로운 변수에 복사할 때는 copy()를 사용해 주세요

In [ ]:
copy_df = df.copy() #카피된 자료를 편집시키면서 여러가지를 해도 원본데이터는 편하지 않는다.

2 . row, column 추가 및 삭제

2-1. row의 추가

In [ ]:
df.head()

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745
3     뷔  방탄소년단  빅히트  남자  1995-12-30  178.0  AB   8073501
4    화사    마마무  RBW  여자  1995-07-23  162.1   A   7650928

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.csv')
import pandas as pd

dictionary 형태의 데이터를 만들어 준 다음 append()함수를 사용하여 데이터를 추가할 수 있습니다. 반드시, ignore_index = True옵션을 같이 추가해 주셔야 에러가 안납니다.

In [ ]:
df = df.append({'이름' : '테디', '그룹' : '테디그룹', '소속사' :'끝내주는소속사', '성별' : '남자', '생년월일' : '1970-01-01', '키': 195.0, '혈액형' : '0', '브랜드평판지수': 12345678 },ignore_index = True)
df

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442
15    테디   테디그룹  끝내주는소속사  남자  1970-01-01  195.0   0  12345678

In [ ]:
df.tail()

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
11  차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12   백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13   JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14   슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442
15   테디   테디그룹  끝내주는소속사  남자  1970-01-01  195.0   0  12345678

2- 2. column 추가

column 추가는 매우 쉽습니다. 단순히 새로운 column을 만들고 값을 대입해주면, 자동으로 생성됩니다.

In [ ]:
df.head()

In [ ]:
df['국적'] = '대한민국'
df.head()

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수    국적
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260  대한민국
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947  대한민국
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745  대한민국
3     뷔  방탄소년단  빅히트  남자  1995-12-30  178.0  AB   8073501  대한민국
4    화사    마마무  RBW  여자  1995-07-23  162.1   A   7650928  대한민국

전부 동일한 값이 추가 되면서, column도 추가된 것을 보실 수 있습니다.

만약, 값을 변경하고 싶다면, loc함수를 활용해서 변경해 줄 수 있겠죠?

In [ ]:
df.loc[df['이름'] == '지드래곤', '국적'] = 'korea'
df.head() 

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수     국적
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260   대한민국
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947  korea
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745   대한민국
3     뷔  방탄소년단  빅히트  남자  1995-12-30  178.0  AB   8073501   대한민국
4    화사    마마무  RBW  여자  1995-07-23  162.1   A   7650928   대한민국

3. 통계값 다루기

In [ ]:
from google.colab import drive
drive.mount('/content/mydrive')

Mounted at /content/mydrive


In [ ]:
import pandas as pd
df= pd.read_csv('https://bit.ly/ds-korean-idol')

통계값은 data type이 float나 int형 인 column을 다룹니다.

In [ ]:
df.describe() #count 해당 컬럼의 개수(row) / mean 평균 / std 표준편차 / min 최소값 / 각 퍼센트대에 대한 값(지표) / max 최대값

키       브랜드평판지수
count   13.000000  1.500000e+01
mean   175.792308  5.655856e+06
std      5.820576  2.539068e+06
min    162.100000  2.925442e+06
25%    174.000000  3.712344e+06
50%    177.000000  4.668615e+06
75%    179.200000  7.862214e+06
max    183.000000  1.052326e+07

3-1. min(최소값), max(최대값)

통계값은 data type이 float나 int형 인 column을 다룹니다

In [ ]:
df['키'].min()

162.1

In [ ]:
df['키'].max()

183.0

3-2. sum(합계), mean(평균)

In [ ]:
df['키'].sum()

In [ ]:
df['키'].mean()

3-3. var, variance(분산)/ std, standard deviation(표준편차)

  - 분산과 표준편차는 데이터가 평균으로부터 얼마나 떨어져 있는지 정도를 나타냅니다.
  - 분산은(데이터- 평균)**2을 모두 합한 값 / n
  - 표준편차는 분산의 루트

In [ ]:
import numpy as np

In [ ]:
data_01 = np.array([1, 3, 5, 7, 9])
data_02 = np.array([3, 4, 5, 6, 7])

In [ ]:
data_01.mean()

5.0

In [ ]:
data_02.mean() #둘의 평균이 같으면 분산 도 같을까?

5.0

In [ ]:
data_01.var(), data_02.var() # 분산 값이 크다 = 데이터가 더 넓게 퍼저 있다.

(8.0, 2.0)

In [ ]:
np.sqrt(data_01.var()), np.sqrt(data_02.var()) #분산에 루트값을 씌우면 표준편차데이터가 나온다(sqrt = 제곱근 = 루트)

(2.8284271247461903, 1.4142135623730951)

In [ ]:
data_01.std(), data_02.std()

(2.8284271247461903, 1.4142135623730951)

주로 분산보다 표준편차(std)를 많이 사용하며, 데이터가 평균으로 부터 얼마나 퍼져있는지 정도를 나타내는 지표라고 이해하시면 됩니다.

In [ ]:
df['키'].var()

33.879102564102595

In [ ]:
df['키'].std()

5.820575793175672

3-4. 갯수를 세는 count

In [ ]:
df['키'].count()

13

3-5. 중앙값

In [ ]:
df['키'].median()

177.0

3-6. 최빈값(mode)

In [ ]:
df['키'].mode()

0    178.0
dtype: float64

4. 피벗테이블(pivot_table)

  - 피벗테이블은 엑셀의 피벗테이블과 동일합니다.
  - 데이터 열 중에서 두 개의 열을 각각 행 인덱스, 열 인덱스로 사용하여 데이터를 조화하여 펼쳐놓은 것을 의미합니다.
  - 왼쪽에 나타나는 인덱스를 행 인덱스, 상단에 나타나는 인덱스를 열 인덱스라고 부릅니다.

In [ ]:
df.head()

In [ ]:
pd.pivot_table(df, index = '소속사', columns = '혈액형', values = '키') # 겹치는데이터는 평균값이 나옴

혈액형          A     AB      B       O
소속사                                 
RBW      162.1    NaN    NaN     NaN
YG       177.0    NaN    NaN     NaN
빅히트      175.8  178.0    NaN  176.60
스타크루이엔티  167.1    NaN    NaN     NaN
커넥트      180.0    NaN    NaN     NaN
판타지오       NaN    NaN  183.0     NaN
플레디스       NaN  175.0    NaN  179.15

In [ ]:
index는 행 인덱스
columns는 열 인덱스
values는 조회하고 싶은 값
즉 행인덱스와 열인덱스는 내가 분류하고싶은 것 이고 values는 내가 확인하고 싶은 값을 의미한다.

In [ ]:
# aggfunc에는 추가 계산 옵션 (np.sum, np.mean) - 기본값은 평균
import numpy as np
pd.pivot_table(df, index = '그룹', columns = '혈액형', values = '브랜드평판지수', aggfunc = np.sum)

혈액형             A         AB          B          O
그룹                                                
뉴이스트          NaN  3301654.0        NaN  8263929.0
마마무     7650928.0        NaN        NaN        NaN
방탄소년단  15731595.0  8073501.0        NaN  7495750.0
빅뱅      9916947.0        NaN        NaN        NaN
소녀시대    3918661.0        NaN        NaN        NaN
아스트로          NaN        NaN  3506027.0        NaN
아이들           NaN        NaN  4668615.0        NaN
핫샷      4036489.0        NaN        NaN        NaN

5. GroupBy (그룹으로 묶어 보기)

  - groupby는 데이터를 그룹으로 묶어 분석할 때 활용합니다.
  - 소속사별 키의 평균, 성별 키의 평균 등 특정, 그룹별 통계 및 데이터 성질을 확인하고자 할 때 활용합니다.

In [99]:
df.head()

지역명           규모구분    연도   월  분양가격
0  서울             전체  2015  10  5841
1  서울          60㎡이하  2015  10  5652
2  서울    60㎡초과 85㎡이하  2015  10  5882
3  서울   85㎡초과 102㎡이하  2015  10  5721
4  서울         102㎡초과  2015  10  5879

In [ ]:
df.groupby('소속사') # 묶어둔 상태 이다.  이제 취하고자 하는 통계 함수를 넣어주면된다.

KeyError: ignored

groupby와 함께
  
  - count() - 갯수
  - sum() - 합계
  - mean() - 평균
  - var() - 분산
  - std() - 표준편차
  - min()/max() - 최소값,최대값


In [ ]:
df.groupby('소속사').count()

이름  그룹  성별  생년월일  키  혈액형  브랜드평판지수
소속사                                       
RBW       1   1   1     1  1    1        1
SM        1   1   1     1  0    1        1
YG        1   1   1     1  1    1        1
빅히트       5   5   5     5  5    5        5
스타크루이엔티   1   1   1     1  1    1        1
커넥트       1   0   1     1  1    1        1
큐브        1   1   1     1  0    1        1
판타지오      1   1   1     1  1    1        1
플레디스      3   3   3     3  3    3        3

산술 통계는 자동으로 산술통계가 가능한 열만 출력됩니다.

In [ ]:
df.groupby('그룹').mean()

키       브랜드평판지수
그룹                             
뉴이스트   177.766667  3.855194e+06
마마무    162.100000  7.650928e+06
방탄소년단  176.560000  6.260169e+06
빅뱅     177.000000  9.916947e+06
소녀시대          NaN  3.918661e+06
아스트로   183.000000  3.506027e+06
아이들           NaN  4.668615e+06
핫샷     167.100000  4.036489e+06

In [ ]:
df.groupby('성별').sum()

키   브랜드평판지수
성별                  
남자  2123.2  68599637
여자   162.1  16238204

특정 열만 출력하고 싶다면?

In [ ]:
df.groupby('혈액형')['키'].mean() 

혈액형
A     172.966667
AB    176.500000
B     183.000000
O     177.875000
Name: 키, dtype: float64

6. Multi-Index(복합 인덱스)

6-1. Multi-Index 적용

행 인덱스를 복합적으로 구성하고 싶은 경우 인덱스를 리스트로 만들어 줍니다.

In [ ]:
df.groupby(['혈액형','성별']).mean() # 혈액형 별로 분류하고 다시한번 성별로 분류해준다.

키     브랜드평판지수
혈액형 성별                     
A   남자  175.140  7591755.20
    여자  162.100  5784794.50
AB  남자  176.500  5687577.50
B   남자  183.000  3506027.00
    여자      NaN  4668615.00
O   남자  177.875  3939919.75

6-2. Multi-Index 데이터 프레임을 피벗 테이블로 변환

 Multi-Index로 된 데이터 프레임을 피벗 테이블 형태로 다시 변환해 줄 수 있습니다.

In [100]:
df2 = df.groupby(['혈액형','성별']).mean()


KeyError: ignored

In [ ]:
df2.unstack('혈액형')

키                           브랜드평판지수                                  
혈액형       A     AB      B        O          A         AB          B           O
성별                                                                             
남자   175.14  176.5  183.0  177.875  7591755.2  5687577.5  3506027.0  3939919.75
여자   162.10    NaN    NaN      NaN  5784794.5        NaN  4668615.0         NaN

In [ ]:
df2.unstack('성별') 

키            브랜드평판지수           
성별        남자     여자          남자         여자
혈액형                                       
A    175.140  162.1  7591755.20  5784794.5
AB   176.500    NaN  5687577.50        NaN
B    183.000    NaN  3506027.00  4668615.0
O    177.875    NaN  3939919.75        NaN

6-3. 인덱스 초기화(reset_index)

reset_index()는 Multi-index로 구성된 데이터 프레임의 인덱스를 초기화해 줍니다.

In [ ]:
df2

키     브랜드평판지수
혈액형 성별                     
A   남자  175.140  7591755.20
    여자  162.100  5784794.50
AB  남자  176.500  5687577.50
B   남자  183.000  3506027.00
    여자      NaN  4668615.00
O   남자  177.875  3939919.75

In [ ]:
df2.reset_index() # 컬럼을 인덱스로 바껴져있는걸 일반 데이터 프레임으로 초기화 시켜준다.
#groupby를 사용해서 데이터 통계를 확인하고 따로 요약된 데이터프레임을 가지고 다른곳에 다른방법으로 쓰기위해 
#다시 되돌릴때 사용한다.

혈액형  성별        키     브랜드평판지수
0   A  남자  175.140  7591755.20
1   A  여자  162.100  5784794.50
2  AB  남자  176.500  5687577.50
3   B  남자  183.000  3506027.00
4   B  여자      NaN  4668615.00
5   O  남자  177.875  3939919.75

In [ ]:
pd.read_csv('/content/mydrive/MyDrive/pandas/02. Part1.[파이썬 필수 스킬] 데이터분석의 시작, 전처리 라이브러리/data/korean-idol.csv')

# 전처리
###1. 결측값을 채워주는 fillna
fillna(): na값에 대하여 fill해주는 함수입니다.



```
# 코드로 형식 지정됨
```



In [ ]:
from google.colab import drive
drive.mount('/content/mydrive')

In [ ]:
import pandas as pd
df =pd.read_csv('https://bit.ly/ds-korean-idol')

In [ ]:
df.info()

키에 2개의 데이터가 누락, 그룹에 1개의 데이터가 누락된 것을 확인 할 수 있습니다.

먼저, 키가없는 사람은 존재할 수 없기 때문에 데이터가 잘 못 되어 있다는 것을 표기 하기 위하여, 누락된 데이터를 -1로 채워 보도록 하겠습니다.

In [ ]:
df['키']

In [ ]:
df['키'].fillna(-1)  # -1 말고 최빈값, 평균값 등등으로 설정 할 수  있음

NaN값이 -1 로 바뀐 것을 확인할 수 있습니다.

하지만, 키의NaN 값을 채워준 다음 유지시키려면 inplace = True 옵션을 주거나, fillna로 채워 준 값을 다시 대입해 주어야 합니다.

In [ ]:
df2 = df.copy()
df2

In [ ]:
df2['키'].fillna(-1, inplace = True)
df2

In [ ]:
이젠 유지가 됩니다. 하지만, 저는 fillna로 채워준 값을 다시 대입 해 주는 방식을 추천 드립니다.

In [ ]:
df2 = df.copy()
df2

In [ ]:
df2['키'] = df2['키'].fillna(-1)
df2['키']

2. 빈 값(NaN)이 있는 행을 제거

  1번 STEP에서 fillna 메소드를 활용하여 값을 채워 주었습니다.
  이번에는 빈값이 있는 행을 제거하는 방법을 알아보도록 하겠습니다.

In [ ]:
df.info()

In [ ]:
df

In [ ]:
df.dropna()

dropna() 는 몇가지 옵션을 추가할 수 있습니다.





 2-1. axis(열/행 을 드랍)

  axis = 0은 행을 드랍합니다.

In [ ]:
df.dropna(axis = 0)

In [ ]:
df.dropna(axis = 1) # NaN값이 있는 그룹과 키 가 날라감

2-2. how옵션 - 'any': 한개라도 있는 경우 드랍, 'all'은 모두 NaN인 경우 드랍( 위에 까지는 하나라도 있는경우 다날렸지만 세밀하게 컨트롤하기 위함)

In [ ]:
df.dropna(axis =0, how= 'any')

In [ ]:
df.dropna(axis =0, how= 'all')

In [ ]:
import numpy as np
df.iloc[10] = np.nan

In [ ]:
df

In [ ]:
df.dropna(axis =0, how= 'all')

3. 중복된 값 제거 (drop_duplicates)

In [ ]:
df = pd.read_csv('http://bit.ly/ds-korean-idol')

In [ ]:
df

3-1. column의 중복값 제거

In [ ]:
df['키']

In [ ]:
df['키'].drop_duplicates() # 기본적으로 첫번째 값 유지시키고 그다음 부터 중복값을 제거 함.

keep 옵션으로 유지하고 싶은 데이터를 선택할 수 있습니다. keep: 'first'/ 'last'

In [ ]:
df['키'].drop_duplicates(keep = 'last')


3-2. 행 전체 제거

In [ ]:
df.drop_duplicates('그룹') #데이터 프레임에서 그룹에서 겹치는 애들이 있으면 행전체를 drop

In [ ]:
df.drop_duplicates('그룹', keep = 'last')

4. Drop - column/row 제거하기(행과 열 제거하기)

In [ ]:
df.head()

4-1. column제거하기

drop()을 활용하여 column을 제거할 수 있습니다. column을 제거할 때는 axis = 1옵션을 줍니다.

In [ ]:
df.drop('그룹',axis = 1)

복수의 column를 제거하고자 할 때는 list로 지정합니다.

In [ ]:
df.drop(['그룹','소속사'], axis = 1)

4-2. row제거하기

drop()을 활용하여 row를 제거할 수 있습니다. row를 제거할 때는 제거하고자하는 index와 axis = 0옵션을 줍니다.

In [ ]:
df.drop(3, axis = 0)

복수의 row를 제거하고자 할 때는 list로 지정합니다.

In [ ]:
df.drop([3, 5], axis = 0)

1. DataFrame 합치기 (concat)

ex) 매출액 데이터가 A회사 B회사 가 있는데 합쳐서 보고싶을때 사용하면 된다.

1-1. Row 기준 합치기

In [2]:
import pandas as pd

df = pd.read_csv('https://bit.ly/ds-korean-idol')
df2 = pd.read_csv('https://bit.ly/ds-korean-idol-2')

In [ ]:
df.head()

In [ ]:
df_copy = df.copy()
df

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

row 에 합칠 때는 pd.concat에 합칠 데이터 프레임을 list로 합쳐줍니다. row기준으로 합칠 때는 sort = False옵션을 주어 순서가 유지되도록 합니다.

In [ ]:
pd.concat([df, df_copy], sort = False)

그런데, index가 꼬였습니다. reset_index()로 인덱스를 초기화 해줄 수 있습니다.

In [ ]:
df_concat = pd.concat([df, df_copy], sort = False)
df_concat

In [ ]:
df_concat.reset_index() 

하지만, index라는 column이 추가 됩니다. 그럴때는 drop = True 옵션으로 새로 index column이 생성되지 않도록 만들어 줍니다.

In [ ]:
df_concat.reset_index(drop = True) 

1- 2. column 기준으로 합치기

이번에는 column 기준으로 횡 방향으로 합치도록 하겠습니다.

In [ ]:
df

In [ ]:
df2

column을 기준으로 합치고자 할 때는 axis = 1옵션을 부여합니다

In [ ]:
pd.concat([df, df2], axis = 1)


행의 개수가 맞지 않는 상태에서 어떻게 될까

In [ ]:
df3 = df2.drop([3, 5])
df3

In [ ]:
pd.concat([df,df3], axis = 1)

2. DataFrame 병합하기 (merge)

  이전에 봤던 concat과 merge는 단순 합치는 목적과 특정 기준(index)으로 합치느냐에 따라 용도가 다릅니다.
  - concat : row 나 column기준으로 단순하게 이어 붙히기
  - merge : 특정 고유한 키(unique id) 값을 기준으로 병합하기

 







In [ ]:
df.head()

In [ ]:
df2.head()

df와 df2는 '이름' 이라는 column이 겹칩니다.

따라서, 우리는 '이름'을 기준으로 두 DataFrame을 병합할 수 있습니다.

pd.merge(left, right, on = '기준column', how = 'left')
  
  - left와 right는 병합할 두 DataFrame을 대입합니다.
  - on 에는 병합의 기준이 되는 column을 넣어 줍니다.
  - how 에는 'left', 'right', 'inner', 'outer' 라는 4가지의 병합 방식중 한가지를 택합니다.
   

2-1. left, right 방식

  먼저, 다음과 같이 df2 에서 5개의 데이터를 고의적으로 drop해주겠습니다.

In [ ]:
df2

In [ ]:
df_right = df2.drop([1, 3, 5, 7, 9], axis = 0)
df_right = df_right.reset_index(drop=True)
df_right

In [ ]:
df

In [ ]:
pd.concat([df, df_right], axis = 1)  
 # df와 df_right를 concat을 했을경우 이름이 전혀 맞지 않는다
 # (위에서drop으로 1,3,5,7,9,를 빼주고 index를 reset했으므로 당연히 순서가 맞지않음) 설명을 위한 설정
 # 지금 하려고 하는것이 적절하게 이름을 알아서 맞춰서 붙히게 하는것

병합할 2개의 DataFrame의 행의 갯수가 맞지 않습니다. 실 사례에서도 충분히 있을 법한 일입니다.

'left'옵션을 부여하면, left DataFrame에 키 값이 존재하면 해당 데이터를 유지하고, 병합한 right DataFrame의 값은 NaN이 대입 됩니다.

In [ ]:
pd.merge(df,df_right, on ='이름', how = 'left')  

반대로 'right' 옵션을 부여하면 right DataFrame을 기준으로 병합하게 됩니다.


현재, left Data Frame이 더 많은 데이터를 보유하고 있으니, right를 기준으로 병합하면 DataFrame 사이즈가 줄어들겠네요.

In [ ]:
pd.merge(df,df_right, on ='이름', how = 'right')  
# 즉 right 데이터 프레임에 없는 이름은 자동으로 drop이 된 모양이다

위의 예제에서, right를 기준으로 병합했기 때문에 DataFrame 사이즈가 줄어들었습니다.

2-2. inner, outer 방식
  
  - inner 방식은 두 DataFrame에 모두 키 값이 존재하는 경우만 병합합니다.
  - outer 방식은 하나의 DataFrame에 키 값이 존재하는 경우 모두 병합합니다. 
  - outer 방식에서는 없는 값은 NaN으로 대입됩니다.

In [ ]:
df

In [ ]:
df_right

In [ ]:
pd.merge(df, df_right, on = '이름', how = 'inner')
# 교집합

In [ ]:
pd.merge(df, df_right, on = '이름', how = 'outer')
#합집합

2-3. column명은 다르지만, 동일한 성질의 데이터 인 경우?
(df_right의 데이터가 만약 이름이 아닌 성함일경우?)

In [ ]:
df

In [ ]:
df_right

In [ ]:
df_right.columns = ['성함', '연봉', '가족수']

In [ ]:
df_right

df와 df_right를 병합하려고 했더니, df에서는  '이름', df_right에서는 '성함' 으로 표기되어 기준이 되는 column을 지정할 수 없습니다.

이럴 때는 left_on, right_on 옵션을 사용합니다.

In [ ]:
pd.merge(df, df_right, left_on= '이름', right_on = '성함', how = 'outer')
#pd.merge(df, df_right, left_on= '이름', right_on = '성함', how = 'inner')

1. Series의 Type

1-1. type 확인하기

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   이름       15 non-null     object 
 1   그룹       14 non-null     object 
 2   소속사      15 non-null     object 
 3   성별       15 non-null     object 
 4   생년월일     15 non-null     object 
 5   키        13 non-null     float64
 6   혈액형      15 non-null     object 
 7   브랜드평판지수  15 non-null     int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 1.1+ KB


이전에 배운 info를 찍어보면, 우측에 type이 같이 찍히는 것을 확인하실 수 있습니다.

  - object : 일반 문자열 타입
  - float : 실수
  - int : 정수
  - category : 카테고리
  - datetime : 시간

1-2. type 변환하기

  - type 변환을 위해서는 astype이라는 메소드를 사용합니다.

키 column 은 현재 float로 명시되어 있습니다.

정수형인 int type으로 변경해 보겠습니다.

In [ ]:
df['키'].dtypes

dtype('float64')

In [ ]:
df['키'].astype(int)

ValueError: ignored

에러가 발생했습니다... NaN값이 들어있기 때문에 변경이 안됩니다.

이럴 때는 fillna로 빈값을 임의로 채워주겠습니다.

In [ ]:
df['키']= df ['키'].fillna(-1)

In [ ]:
df['키']

0     173.6
1     177.0
2     180.0
3     178.0
4     162.1
5     178.0
6     182.3
7      -1.0
8     179.2
9     167.1
10     -1.0
11    183.0
12    175.0
13    176.0
14    174.0
Name: 키, dtype: float64

In [ ]:
df['키'].astype(int)

0     173
1     177
2     180
3     178
4     162
5     178
6     182
7      -1
8     179
9     167
10     -1
11    183
12    175
13    176
14    174
Name: 키, dtype: int64

int형으로 타입을 바꾸니 float -> int로 변경 된 것을 볼 수 있습니다.

1-3. 날짜 변환하기 (datetime 타입)

  - 날짜를 변환하기 위해서는 판다스 메소드인 to_datetime이라는 메소드를 사용합니다.

현재 날짜 column은 dtype: 이 object, 즉 문자열 타입으로 되어 있습니다.

In [ ]:
df['생년월일'] # 시간관련 데이터 타입은 datetime으로 변환을 할 수 있다.

0     1995-10-13
1     1988-08-18
2     1996-12-10
3     1995-12-30
4     1995-07-23
5     1997-09-01
6     1995-08-09
7     1998-08-26
8     1992-12-04
9     1994-03-22
10    1989-03-09
11    1997-03-30
12    1995-07-21
13    1995-06-08
14    1993-03-09
Name: 생년월일, dtype: object

datetime타입으로 바꾸어 볼까요?

In [ ]:
pd.to_datetime(df['생년월일'])

0    1995-10-13
1    1988-08-18
2    1996-12-10
3    1995-12-30
4    1995-07-23
5    1997-09-01
6    1995-08-09
7    1998-08-26
8    1992-12-04
9    1994-03-22
10   1989-03-09
11   1997-03-30
12   1995-07-21
13   1995-06-08
14   1993-03-09
Name: 생년월일, dtype: datetime64[ns]

dtype이 datetime으로 변환된 것을 확인 하셨나요?

하지만, 이렇게 해준다고 해서 우리의 DataFrame에는 반영되어 있지 않습니다.

변환된 것을 df['날짜']column에 다시 대입을 해줘야 정상적으로 변환된 값이 들어 갑니다.

In [ ]:
df['생년월일'] = pd.to_datetime(df['생년월일'])
df['생년월일']

0    1995-10-13
1    1988-08-18
2    1996-12-10
3    1995-12-30
4    1995-07-23
5    1997-09-01
6    1995-08-09
7    1998-08-26
8    1992-12-04
9    1994-03-22
10   1989-03-09
11   1997-03-30
12   1995-07-21
13   1995-06-08
14   1993-03-09
Name: 생년월일, dtype: datetime64[ns]

In [ ]:
datetime 타입으로 우리가 변환해준 중요한 이유가 있습니다!

  - 매우 손쉽게 월, 일, 요일 등등의 날짜 정보를 세부적으로 추출해낼 수 있습니다.
  - datetime의 약어인 'dt'에는 다양한 정보들을 제공해 줍니다.

In [ ]:
df['생년월일'].dt

In [ ]:
df['생년월일'].dt.year

0     1995
1     1988
2     1996
3     1995
4     1995
5     1997
6     1995
7     1998
8     1992
9     1994
10    1989
11    1997
12    1995
13    1995
14    1993
Name: 생년월일, dtype: int64

In [ ]:
df['생년월일'].dt.month

0     10
1      8
2     12
3     12
4      7
5      9
6      8
7      8
8     12
9      3
10     3
11     3
12     7
13     6
14     3
Name: 생년월일, dtype: int64

In [ ]:
df['생년월일'].dt.day

0     13
1     18
2     10
3     30
4     23
5      1
6      9
7     26
8      4
9     22
10     9
11    30
12    21
13     8
14     9
Name: 생년월일, dtype: int64

In [ ]:
# df['생년월일'].dt.hour
# df['생년월일'].dt.second
# df['생년월일'].dt.minutes
df['생년월일'].dt.dayofweek #태어난 날짜가 무슨요일인지
#월요일 : 0 , 화요일 : 1, 수요일 : 2,.......일요일 : 6

0     4
1     3
2     1
3     5
4     6
5     0
6     2
7     2
8     4
9     1
10    3
11    6
12    4
13    3
14    1
Name: 생년월일, dtype: int64

In [ ]:
df['생년월일'].dt.weekofyear #태어난 날이 1년 기준으로 몇주차 되는 주인지

In [ ]:
df.head()

이름     그룹  소속사  성별       생년월일      키 혈액형   브랜드평판지수
0    지민  방탄소년단  빅히트  남자 1995-10-13  173.6   A  10523260
1  지드래곤     빅뱅   YG  남자 1988-08-18  177.0   A   9916947
2  강다니엘    NaN  커넥트  남자 1996-12-10  180.0   A   8273745
3     뷔  방탄소년단  빅히트  남자 1995-12-30  178.0  AB   8073501
4    화사    마마무  RBW  여자 1995-07-23  162.1   A   7650928

In [ ]:
df['생일_월'] = df['생년월일'].dt.month
df['생일_년'] = df['생년월일'].dt.year
df['생일_일'] = df['생년월일'].dt.day

In [ ]:
df.head()

이름     그룹  소속사  성별       생년월일      키 혈액형   브랜드평판지수  생일_월  생일_년  생일_일
0    지민  방탄소년단  빅히트  남자 1995-10-13  173.6   A  10523260    10  1995    13
1  지드래곤     빅뱅   YG  남자 1988-08-18  177.0   A   9916947     8  1988    18
2  강다니엘    NaN  커넥트  남자 1996-12-10  180.0   A   8273745    12  1996    10
3     뷔  방탄소년단  빅히트  남자 1995-12-30  178.0  AB   8073501    12  1995    30
4    화사    마마무  RBW  여자 1995-07-23  162.1   A   7650928     7  1995    23

1. apply(기본적으로 행)

In [3]:
df

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

apply는 Series나 DataFrame에 좀 더 구체적인 로직을 적용하고 싶은 경우 활용합니다.

  - apply 를 적용하기 위해서는 함수가 먼저 정의되어야합니다.
  - apply 는 정의한 로직 함수를 인자로 넘겨줍니다.

1-1. (목표) 남자/여자의 문자열 데이터로 구성된 '성별' column을 1/ 0 으로 바꿔보세요

남자: 1 여자 : 0 기타 : -1

**[주의] 반드시 return값이 존재하여야 합니다.**

In [4]:
df

이름     그룹      소속사  성별        생년월일      키 혈액형   브랜드평판지수
0     지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260
1   지드래곤     빅뱅       YG  남자  1988-08-18  177.0   A   9916947
2   강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
3      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
4     화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
5     정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   5208335
6     민현   뉴이스트     플레디스  남자  1995-08-09  182.3   O   4989792
7     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
8      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
10    태연   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
11   차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
12    백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
13    JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
14    슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   2925442

In [5]:
# 먼저 apply에 활용할 함수를 정의합니다.
def male_or_female(x):
  if x == '남자':
    return 1 
  elif x == '여자':
    return 0

In [6]:
df['성별'].apply(male_or_female)

0     1
1     1
2     1
3     1
4     0
5     1
6     1
7     0
8     1
9     1
10    0
11    1
12    1
13    1
14    1
Name: 성별, dtype: int64

1-2. (목표)cm당 브랜드 평판지수를 구해보세요(브랜드평판지수/키)

키: 178/ 브랜드평판지수 : 99000
값 : 99000 / 178

In [8]:
# 먼저 apply에 활용할 함수를 정의합니다.
def cm_to_brand(df):
  value = df['브랜드평판지수'] / df['키']
  return value

In [11]:
df.apply(cm_to_brand, axis =1 ) 
#브랜드평판지수와 키 두개가 필요해서 단일 시리즈가 아닌 데이터프레임 자체가 필요
# 기본적으로 행을 넘겨주기땜에 컬럼을 넘겨줄때 axis = 1 을 넣어준다.

0     60617.857143
1     56027.949153
2     45965.250000
3     45356.747191
4     47198.815546
5     29260.308989
6     27371.321997
7              NaN
8     25503.950893
9     24156.128067
10             NaN
11    19158.617486
12    18866.594286
13    18603.051136
14    16812.885057
dtype: float64

2. lambda 함수의 적용

  - lambda는 1줄로 작성하는 간단 함수식입니다.
  - return을 별도로 명기하지 않습니다.

In [ ]:
df['성별'].apply(lambda x: 1 if x == '남자' else 0) #lambda x 란 x로 값이 넘어오는 인자를 말함 즉 df['성별']의 값이 x로 하나하나 들어간다.

0     1
1     1
2     1
3     1
4     0
5     1
6     1
7     0
8     1
9     1
10    0
11    1
12    1
13    1
14    1
Name: 성별, dtype: int64

하지만, 위에것을 보면 그렇게 직관적이지 않아서 함수식 사용해도된다

실제로는 간단한 계산식을 적용하려는 경우에 많이 사용합니다.

In [ ]:
df['키'].apply(lambda x: x / 2)

0     86.80
1     88.50
2     90.00
3     89.00
4     81.05
5     89.00
6     91.15
7       NaN
8     89.60
9     83.55
10      NaN
11    91.50
12    87.50
13    88.00
14    87.00
Name: 키, dtype: float64

In [ ]:
df['키'].apply(lambda x: x ** 2)

0     30136.96
1     31329.00
2     32400.00
3     31684.00
4     26276.41
5     31684.00
6     33233.29
7          NaN
8     32112.64
9     27922.41
10         NaN
11    33489.00
12    30625.00
13    30976.00
14    30276.00
Name: 키, dtype: float64

apply에 함수식을 만들어서 적용해주는 것과 동일하기 때문에, 복잡한 조건식은 함수로 간단한 계산식은 lambda로 적용할 수 있습니다.

3. map- 값을 매핑!

In [ ]:
my_map = {'남자' : 1, '여자' : 0}
#map은 반드시 사전으로 정의

In [ ]:
df['성별'].map(my_map)

0     1
1     1
2     1
3     1
4     0
5     1
6     1
7     0
8     1
9     1
10    0
11    1
12    1
13    1
14    1
Name: 성별, dtype: int64

1. Column과 Column 간 연산(+, -, *, /, %)

In [12]:
import pandas as pd
import numpy as np
df = pd.DataFrame({'통계': [60, 70, 80, 85, 75], '미술': [50, 55, 80, 100, 95], '체육': [70, 65, 50, 95, 100]})
df

통계   미술   체육
0  60   50   70
1  70   55   65
2  80   80   50
3  85  100   95
4  75   95  100

In [13]:
df['통계'] + df['미술'] + df['체육']

0    180
1    190
2    210
3    280
4    270
dtype: int64

In [14]:
df['통계'] - df['미술']

0    10
1    15
2     0
3   -15
4   -20
dtype: int64

In [15]:
df['통계'] * df['미술']

0    3000
1    3850
2    6400
3    8500
4    7125
dtype: int64

In [16]:
df['통계'] / df['미술']

0    1.200000
1    1.272727
2    1.000000
3    0.850000
4    0.789474
dtype: float64

In [17]:
df['통계'] % df['미술']

0    10
1    15
2     0
3    85
4    75
dtype: int64

2. Column과 숫자 간 연산

In [ ]:
df['통계'] + 10 
# df['통계'] - 10 

0    70
1    80
2    90
3    95
4    85
Name: 통계, dtype: int64

3. 복합연산

In [18]:
df['통계미술합계'] = df['통계'] + df['미술'] + 10
df

통계   미술   체육  통계미술합계
0  60   50   70     120
1  70   55   65     135
2  80   80   50     170
3  85  100   95     195
4  75   95  100     180

In [21]:
df.mean(axis = 1)
# df.mean(axis = 0)

0     75.00
1     81.25
2     95.00
3    118.75
4    112.50
dtype: float64

4. mean(), sum()을 axis 기준으로 연산

In [22]:
df = pd.DataFrame({'통계': [60, 70, 80, 85, 75], '미술': [50, 55, 80, 100, 95], '체육': [70, 65, 50, 95, 100]})
df

통계   미술   체육
0  60   50   70
1  70   55   65
2  80   80   50
3  85  100   95
4  75   95  100

In [24]:
df.sum(axis = 0)
#df.sum(axis = 1)

통계    370
미술    380
체육    380
dtype: int64

In [25]:
df.mean(axis = 0)
#df.mean(axis = 1)

통계    74.0
미술    76.0
체육    76.0
dtype: float64

5. NaN값이 존재할 경우 연산

In [26]:
df = pd.DataFrame({'통계' : [60, 70, np.nan, 85, 75], '미술' : [50, np.nan, 80, 100, 95], '체육' : [70, 65, 50, np.nan, 100]})
df

통계     미술     체육
0  60.0   50.0   70.0
1  70.0    NaN   65.0
2   NaN   80.0   50.0
3  85.0  100.0    NaN
4  75.0   95.0  100.0

In [27]:
df['통계'] / 2

0    30.0
1    35.0
2     NaN
3    42.5
4    37.5
Name: 통계, dtype: float64

In [28]:
1000 / df['통계']

0    16.666667
1    14.285714
2          NaN
3    11.764706
4    13.333333
Name: 통계, dtype: float64

In [29]:
df['통계'] / np.nan

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: 통계, dtype: float64

In [ ]:
np.nan / df['통계']

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: 통계, dtype: float64

6. DataFrame 과 DataFrame 간 연산

In [30]:
df1 = pd.DataFrame({'통계' : [60, 70, 80, 85, 75], '미술' : [50, 55, 80, 100, 95], '체육' : [70, 65, 50, 95, 100]})
df2 = pd.DataFrame({'통계' : ['good', 'good', 'ok', 'good', 'ok'], '미술' : [50, 60, 80, 100, 95], '체육' : [70, 65, 50, 70, 100]})


In [31]:
df1


통계   미술   체육
0  60   50   70
1  70   55   65
2  80   80   50
3  85  100   95
4  75   95  100

In [32]:
df2

통계   미술   체육
0  good   50   70
1  good   60   65
2    ok   80   50
3  good  100   70
4    ok   95  100

문자열이 포함된 DataFrame의 경우

In [ ]:
df1 + df2  #에러가 난다

TypeError: ignored

문자열이 포함된 DataFrame의 브로드캐스팅 연산

In [ ]:
df1 + 10

통계   미술   체육
0  70   60   80
1  80   65   75
2  90   90   60
3  95  110  105
4  85  105  110

In [ ]:
df2 + 10

TypeError: ignored

column의 순서가 바뀌어 있는 경우

In [ ]:
df1 = pd.DataFrame({'통계' : [10,20,30,40,50], '미술' : [60,70,80,90,100]})
df2 = pd.DataFrame({'미술' : [10,20,30,40,50], '통계' : [60,70,80,90,100]})
df1 + df2
# 같은 column끼리 계산 됨

미술   통계
0   70   70
1   90   90
2  110  110
3  130  130
4  150  150

행의 갯수가 다른경우

In [ ]:
df1 = pd.DataFrame({'통계' : [10,20,30,40,50,60], '미술' : [60,70,80,90,100,110]})
df2 = pd.DataFrame({'미술' : [10,20,30,40,50], '통계' : [60,70,80,90,100]})
df1 * df2
# 부족한 곳은 nan값이므로 nan값으로 할당됨

미술      통계
0   600.0   600.0
1  1400.0  1400.0
2  2400.0  2400.0
3  3600.0  3600.0
4  5000.0  5000.0
5     NaN     NaN

1. 데이터 타입별 column 선택 (select_dtypes)

In [33]:
df = pd.read_csv('https://bit.ly/ds-korean-idol')

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   이름       15 non-null     object 
 1   그룹       14 non-null     object 
 2   소속사      15 non-null     object 
 3   성별       15 non-null     object 
 4   생년월일     15 non-null     object 
 5   키        13 non-null     float64
 6   혈액형      15 non-null     object 
 7   브랜드평판지수  15 non-null     int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 1.1+ KB


In [36]:
df.describe()

키       브랜드평판지수
count   13.000000  1.500000e+01
mean   175.792308  5.655856e+06
std      5.820576  2.539068e+06
min    162.100000  2.925442e+06
25%    174.000000  3.712344e+06
50%    177.000000  4.668615e+06
75%    179.200000  7.862214e+06
max    183.000000  1.052326e+07

1-1. 문자열이 있는 column만 선택

In [38]:
df.select_dtypes(include ='object')

이름     그룹      소속사  성별        생년월일 혈액형
0     지민  방탄소년단      빅히트  남자  1995-10-13   A
1   지드래곤     빅뱅       YG  남자  1988-08-18   A
2   강다니엘    NaN      커넥트  남자  1996-12-10   A
3      뷔  방탄소년단      빅히트  남자  1995-12-30  AB
4     화사    마마무      RBW  여자  1995-07-23   A
5     정국  방탄소년단      빅히트  남자  1997-09-01   A
6     민현   뉴이스트     플레디스  남자  1995-08-09   O
7     소연    아이들       큐브  여자  1998-08-26   B
8      진  방탄소년단      빅히트  남자  1992-12-04   O
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22   A
10    태연   소녀시대       SM  여자  1989-03-09   A
11   차은우   아스트로     판타지오  남자  1997-03-30   B
12    백호   뉴이스트     플레디스  남자  1995-07-21  AB
13    JR   뉴이스트     플레디스  남자  1995-06-08   O
14    슈가  방탄소년단      빅히트  남자  1993-03-09   O

In [40]:
df.select_dtypes(exclude ='object')

키   브랜드평판지수
0   173.6  10523260
1   177.0   9916947
2   180.0   8273745
3   178.0   8073501
4   162.1   7650928
5   178.0   5208335
6   182.3   4989792
7     NaN   4668615
8   179.2   4570308
9   167.1   4036489
10    NaN   3918661
11  183.0   3506027
12  175.0   3301654
13  176.0   3274137
14  174.0   2925442

(활용) 문자열이 포함된 DataFrame의 연산으로 발생되는 Error를 피하기 위해서는 int형 데이터만 따로 뽑고 계산하면된다!!

In [41]:
df.select_dtypes(include ='object') + 10 

TypeError: ignored

In [42]:
df.select_dtypes(exclude ='object') + 10

키   브랜드평판지수
0   183.6  10523270
1   187.0   9916957
2   190.0   8273755
3   188.0   8073511
4   172.1   7650938
5   188.0   5208345
6   192.3   4989802
7     NaN   4668625
8   189.2   4570318
9   177.1   4036499
10    NaN   3918671
11  193.0   3506037
12  185.0   3301664
13  186.0   3274147
14  184.0   2925452

In [43]:
df.select_dtypes(include ='float') + 10

키
0   183.6
1   187.0
2   190.0
3   188.0
4   172.1
5   188.0
6   192.3
7     NaN
8   189.2
9   177.1
10    NaN
11  193.0
12  185.0
13  186.0
14  184.0

In [44]:
df.select_dtypes(exclude ='object').columns #컬럼만 return받을 수 있다.

Index(['키', '브랜드평판지수'], dtype='object')

In [45]:
num_cols =df.select_dtypes(exclude ='object').columns
df[num_cols]

키   브랜드평판지수
0   173.6  10523260
1   177.0   9916947
2   180.0   8273745
3   178.0   8073501
4   162.1   7650928
5   178.0   5208335
6   182.3   4989792
7     NaN   4668615
8   179.2   4570308
9   167.1   4036489
10    NaN   3918661
11  183.0   3506027
12  175.0   3301654
13  176.0   3274137
14  174.0   2925442

In [ ]:
obj_cols =df.select_dtypes(include ='object').columns
df[obj_cols]

이름     그룹      소속사  성별        생년월일 혈액형
0     지민  방탄소년단      빅히트  남자  1995-10-13   A
1   지드래곤     빅뱅       YG  남자  1988-08-18   A
2   강다니엘    NaN      커넥트  남자  1996-12-10   A
3      뷔  방탄소년단      빅히트  남자  1995-12-30  AB
4     화사    마마무      RBW  여자  1995-07-23   A
5     정국  방탄소년단      빅히트  남자  1997-09-01   A
6     민현   뉴이스트     플레디스  남자  1995-08-09   O
7     소연    아이들       큐브  여자  1998-08-26   B
8      진  방탄소년단      빅히트  남자  1992-12-04   O
9    하성운     핫샷  스타크루이엔티  남자  1994-03-22   A
10    태연   소녀시대       SM  여자  1989-03-09   A
11   차은우   아스트로     판타지오  남자  1997-03-30   B
12    백호   뉴이스트     플레디스  남자  1995-07-21  AB
13    JR   뉴이스트     플레디스  남자  1995-06-08   O
14    슈가  방탄소년단      빅히트  남자  1993-03-09   O

2. 원핫인코딩(One-hot-encoding)

  - 원핫인코딩은 한개의 요소는 True 그리고 나머지 요소는 False로 만들어 주는 기법입니다.
  - 원핫인코딩을 왜 필요할까요

In [46]:
df = pd.read_csv('https://bit.ly/ds-korean-idol')

In [47]:
df.head()

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745
3     뷔  방탄소년단  빅히트  남자  1995-12-30  178.0  AB   8073501
4    화사    마마무  RBW  여자  1995-07-23  162.1   A   7650928

In [48]:
blood_map = {
    'A' : 0,
    'B' : 1,
    'AB' : 2,
    'O' : 3
}

In [49]:
df['혈액형'].map(blood_map)

0     0
1     0
2     0
3     2
4     0
5     0
6     3
7     1
8     3
9     0
10    0
11    1
12    2
13    3
14    3
Name: 혈액형, dtype: int64

In [50]:
df['혈액형_code'] = df['혈액형'].map(blood_map)

In [51]:
df.head()

이름     그룹  소속사  성별        생년월일      키 혈액형   브랜드평판지수  혈액형_code
0    지민  방탄소년단  빅히트  남자  1995-10-13  173.6   A  10523260         0
1  지드래곤     빅뱅   YG  남자  1988-08-18  177.0   A   9916947         0
2  강다니엘    NaN  커넥트  남자  1996-12-10  180.0   A   8273745         0
3     뷔  방탄소년단  빅히트  남자  1995-12-30  178.0  AB   8073501         2
4    화사    마마무  RBW  여자  1995-07-23  162.1   A   7650928         0

In [52]:
df['혈액형_code'].value_counts()

0    7
3    4
2    2
1    2
Name: 혈액형_code, dtype: int64

  - 우리가 만약 df['혈액형_code']를 머신러닝 알고리즘에 그대로 넣어 데이터를 예측하라고 지시한다면, 컴퓨터는 '혈액형_code'안에서 값들간의 관계를 스스로 형성하게 됩니다.
  - 이 상황에서 만약B형은 1, AB형은 2 라는 값을 가지고 있는데, 컴퓨터는 B형 + AB형 = O형이다 라고 잘못 관계를 맺을 수 있게 됩니다.
  - 따라서, 우리는 4개의 별도의 column을 형성해주고 1개의 column에는 True 나머지는 모두 False를 넣어 줌으로써 A, B, AB, O형의 관계는 모두 독립적이다를 표현해 줍니다.
  - 이를 원핫인코딩이라고합니다.
  

In [53]:
df['혈액형_code']

0     0
1     0
2     0
3     2
4     0
5     0
6     3
7     1
8     3
9     0
10    0
11    1
12    2
13    3
14    3
Name: 혈액형_code, dtype: int64

In [54]:
pd.get_dummies(df['혈액형_code'])
# 하지만 이렇게하면 보기가 불편
#https://devuna.tistory.com/67 참고

0  1  2  3
0   1  0  0  0
1   1  0  0  0
2   1  0  0  0
3   0  0  1  0
4   1  0  0  0
5   1  0  0  0
6   0  0  0  1
7   0  1  0  0
8   0  0  0  1
9   1  0  0  0
10  1  0  0  0
11  0  1  0  0
12  0  0  1  0
13  0  0  0  1
14  0  0  0  1

prefix를 설정하려면?

In [55]:
pd.get_dummies(df['혈액형_code'], prefix='혈액형')

혈액형_0  혈액형_1  혈액형_2  혈액형_3
0       1      0      0      0
1       1      0      0      0
2       1      0      0      0
3       0      0      1      0
4       1      0      0      0
5       1      0      0      0
6       0      0      0      1
7       0      1      0      0
8       0      0      0      1
9       1      0      0      0
10      1      0      0      0
11      0      1      0      0
12      0      0      1      0
13      0      0      0      1
14      0      0      0      1

나중에 머신러닝 알고리즘으로 발전하기 위해서는 꼭 알아야하는 개념입니다! 지금은 원핫인코딩의 개념만 이해하시면 됩니다.

# pandas전처리 실습

In [57]:
import pandas as pd

data.or.kr 에서 제공하는 공공데이터를 활용합니다

민간 아파트 가격동향

## 데이터 프레임 로드

In [58]:
 import pandas as pd
 df = pd.read_csv('https://bit.ly/ds-house-price')

In [59]:
df

지역명               규모구분    연도   월 분양가격(㎡)
0     서울                 전체  2015  10    5841
1     서울         전용면적 60㎡이하  2015  10    5652
2     서울   전용면적 60㎡초과 85㎡이하  2015  10    5882
3     서울  전용면적 85㎡초과 102㎡이하  2015  10    5721
4     서울        전용면적 102㎡초과  2015  10    5879
...   ..                ...   ...  ..     ...
4500  제주                 전체  2020   2    3955
4501  제주         전용면적 60㎡이하  2020   2    4039
4502  제주   전용면적 60㎡초과 85㎡이하  2020   2    3962
4503  제주  전용면적 85㎡초과 102㎡이하  2020   2     NaN
4504  제주        전용면적 102㎡초과  2020   2    3601

[4505 rows x 5 columns]

## 1. column 재정의 (rename)

이전 까지는 df.columns 를 통해 전체 column의 이름을 재정의했지만, 한 개의 column만 이름을 바꾸고 싶을 때는 다음과 같이 실행합니다.

분양가격 column의 이름을 재정의

In [63]:
df = df.rename(columns = {'분양가격(㎡)' : '분양가격'})
df

지역명               규모구분    연도   월  분양가격
0     서울                 전체  2015  10  5841
1     서울         전용면적 60㎡이하  2015  10  5652
2     서울   전용면적 60㎡초과 85㎡이하  2015  10  5882
3     서울  전용면적 85㎡초과 102㎡이하  2015  10  5721
4     서울        전용면적 102㎡초과  2015  10  5879
...   ..                ...   ...  ..   ...
4500  제주                 전체  2020   2  3955
4501  제주         전용면적 60㎡이하  2020   2  4039
4502  제주   전용면적 60㎡초과 85㎡이하  2020   2  3962
4503  제주  전용면적 85㎡초과 102㎡이하  2020   2   NaN
4504  제주        전용면적 102㎡초과  2020   2  3601

[4505 rows x 5 columns]

## 2. 데이터 Overview

2-1. 빈 값과 Data Type 확인하기

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4505 entries, 0 to 4504
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   지역명      4505 non-null   object
 1   규모구분     4505 non-null   object
 2   연도       4505 non-null   int64 
 3   월        4505 non-null   int64 
 4   분양가격(㎡)  4210 non-null   object
dtypes: int64(2), object(3)
memory usage: 176.1+ KB


2-2. 통계값 확인

In [61]:
df.describe()

연도            월
count  4505.000000  4505.000000
mean   2017.452830     6.566038
std       1.311432     3.595519
min    2015.000000     1.000000
25%    2016.000000     3.000000
50%    2017.000000     7.000000
75%    2019.000000    10.000000
max    2020.000000    12.000000

3. 분양가격 column을 int타입으로 변환( 실전 전처리)

In [ ]:
df['분양가격'].astype(int)

ValueError: ignored

실제로 이렇게 처리된 데이터들이 많다...

strip() 을 활용하여 공백이 있는 데이터 공백없애기

[tip] column의 문자열에 strip을 실행하고자 할 때는 str.strip()를 실행해줘야 합니다.

[예시] df['분양가격'].str.strip()

In [64]:
df.loc[df['분양가격'] == '  ']

지역명               규모구분    연도   월 분양가격
28   광주  전용면적 85㎡초과 102㎡이하  2015  10     
29   광주        전용면적 102㎡초과  2015  10     
34   대전        전용면적 102㎡초과  2015  10     
81   제주         전용면적 60㎡이하  2015  10     
113  광주  전용면적 85㎡초과 102㎡이하  2015  11     
114  광주        전용면적 102㎡초과  2015  11     
119  대전        전용면적 102㎡초과  2015  11     
166  제주         전용면적 60㎡이하  2015  11     
198  광주  전용면적 85㎡초과 102㎡이하  2015  12     
199  광주        전용면적 102㎡초과  2015  12     
204  대전        전용면적 102㎡초과  2015  12     
251  제주         전용면적 60㎡이하  2015  12     
283  광주  전용면적 85㎡초과 102㎡이하  2016   1     
284  광주        전용면적 102㎡초과  2016   1     
289  대전        전용면적 102㎡초과  2016   1     
336  제주         전용면적 60㎡이하  2016   1

In [65]:
df['분양가격'] = df['분양가격'].str.strip()
# 컬럼별로 공백제거를 하기 위해서는 .str. 을 사용해 줘야한다.

In [66]:
df.loc[df['분양가격'] == '']

지역명               규모구분    연도   월 분양가격
28   광주  전용면적 85㎡초과 102㎡이하  2015  10     
29   광주        전용면적 102㎡초과  2015  10     
34   대전        전용면적 102㎡초과  2015  10     
81   제주         전용면적 60㎡이하  2015  10     
113  광주  전용면적 85㎡초과 102㎡이하  2015  11     
114  광주        전용면적 102㎡초과  2015  11     
119  대전        전용면적 102㎡초과  2015  11     
166  제주         전용면적 60㎡이하  2015  11     
198  광주  전용면적 85㎡초과 102㎡이하  2015  12     
199  광주        전용면적 102㎡초과  2015  12     
204  대전        전용면적 102㎡초과  2015  12     
251  제주         전용면적 60㎡이하  2015  12     
283  광주  전용면적 85㎡초과 102㎡이하  2016   1     
284  광주        전용면적 102㎡초과  2016   1     
289  대전        전용면적 102㎡초과  2016   1     
336  제주         전용면적 60㎡이하  2016   1

빈 공백이 있는 데이터는 0을 넣어주도록 하겠습니다.

In [67]:
df.loc[df['분양가격'] == '', '분양가격'] = 0
df.loc[df['분양가격'] == 0]

지역명               규모구분    연도   월 분양가격
28   광주  전용면적 85㎡초과 102㎡이하  2015  10    0
29   광주        전용면적 102㎡초과  2015  10    0
34   대전        전용면적 102㎡초과  2015  10    0
81   제주         전용면적 60㎡이하  2015  10    0
113  광주  전용면적 85㎡초과 102㎡이하  2015  11    0
114  광주        전용면적 102㎡초과  2015  11    0
119  대전        전용면적 102㎡초과  2015  11    0
166  제주         전용면적 60㎡이하  2015  11    0
198  광주  전용면적 85㎡초과 102㎡이하  2015  12    0
199  광주        전용면적 102㎡초과  2015  12    0
204  대전        전용면적 102㎡초과  2015  12    0
251  제주         전용면적 60㎡이하  2015  12    0
283  광주  전용면적 85㎡초과 102㎡이하  2016   1    0
284  광주        전용면적 102㎡초과  2016   1    0
289  대전        전용면적 102㎡초과  2016   1    0
336  제주         전용면적 60㎡이하  2016   1    0

In [69]:
df['분양가격'].astype(int) 

ValueError: ignored

이번에는 NaN 값이 말썽이다.

NaN값은 fillna로 채워 주도록 하자

In [70]:
df['분양가격'] = df['분양가격'].fillna(0)

In [71]:
df['분양가격'].astype(int)

ValueError: ignored

또 문제가 생겼다. 콤마가 들어간 데이터 가 존재한다.

In [ ]:
df.loc[df['분양가격'] == '6,657']
# 데이터가 object이므로 찾을때 문자열로 대입('')

지역명 규모구분    연도   월   분양가격
2125  서울   전체  2017  11  6,657

2125행에 ,가 들어간 것이 보이시죠?

콤마를 제거해 봅시다

In [72]:
df['분양가격'] = df['분양가격'].str.replace(',', '')

In [73]:
df.iloc[2125]
#df.loc[df['분양가격'] == '6657']

지역명       서울
규모구분      전체
연도      2017
월         11
분양가격    6657
Name: 2125, dtype: object

6,657 -> 6657로 변환된 것을 보실 수 있네요

In [74]:
df['분양가격'].astype(int)

ValueError: ignored

In [75]:
df['분양가격'] = df['분양가격'].fillna(0)

In [76]:
df['분양가격'].astype(int)

ValueError: ignored

이번에는 '-'가 말썽이네요 

-제거하기

In [77]:
df.loc[df['분양가격'] == '-']

지역명               규모구분    연도  월 분양가격
3683  광주  전용면적 85㎡초과 102㎡이하  2019  5    -
3686  대전         전용면적 60㎡이하  2019  5    -
3688  대전  전용면적 85㎡초과 102㎡이하  2019  5    -
3690  울산                 전체  2019  5    -
3691  울산         전용면적 60㎡이하  2019  5    -
3692  울산   전용면적 60㎡초과 85㎡이하  2019  5    -
3693  울산  전용면적 85㎡초과 102㎡이하  2019  5    -
3694  울산        전용면적 102㎡초과  2019  5    -
3696  세종         전용면적 60㎡이하  2019  5    -

In [79]:
df['분양가격'] = df['분양가격'].str.replace('-', '')

In [81]:
df['분양가격'] = df['분양가격'].fillna(0)

In [82]:
df['분양가격'].astype(int)

ValueError: ignored

빈 값이 생겼다 이것을 0으로 채우기

In [83]:
df.loc[df['분양가격'] == '', '분양가격'] = 0

In [84]:
df['분양가격'] = df['분양가격'].astype(int)

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4505 entries, 0 to 4504
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   지역명     4505 non-null   object
 1   규모구분    4505 non-null   object
 2   연도      4505 non-null   int64 
 3   월       4505 non-null   int64 
 4   분양가격    4505 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 176.1+ KB


규모구분 column에 불필요한 '전용면적' 제거

In [86]:
df

지역명               규모구분    연도   월  분양가격
0     서울                 전체  2015  10  5841
1     서울         전용면적 60㎡이하  2015  10  5652
2     서울   전용면적 60㎡초과 85㎡이하  2015  10  5882
3     서울  전용면적 85㎡초과 102㎡이하  2015  10  5721
4     서울        전용면적 102㎡초과  2015  10  5879
...   ..                ...   ...  ..   ...
4500  제주                 전체  2020   2  3955
4501  제주         전용면적 60㎡이하  2020   2  4039
4502  제주   전용면적 60㎡초과 85㎡이하  2020   2  3962
4503  제주  전용면적 85㎡초과 102㎡이하  2020   2     0
4504  제주        전용면적 102㎡초과  2020   2  3601

[4505 rows x 5 columns]

In [87]:
df['규모구분'] = df['규모구분'].str.replace('전용면적', '')

In [88]:
df['규모구분'].value_counts()
# value_counts() 각각의 value가 몇개 있는지 확인

 60㎡이하           901
 102㎡초과          901
 85㎡초과 102㎡이하    901
전체               901
 60㎡초과 85㎡이하     901
Name: 규모구분, dtype: int64

6. 지역명 별로 평균 분양가격 확인해 봅시다

In [89]:
df.groupby('지역명')['분양가격'].mean()

지역명
강원    2339.807547
경기    4072.667925
경남    2761.275472
경북    2432.128302
광주    2450.728302
대구    3538.920755
대전    2479.135849
부산    3679.920755
서울    7225.762264
세종    2815.098113
울산    1826.101887
인천    3578.433962
전남    2270.177358
전북    2322.060377
제주    2979.407547
충남    2388.324528
충북    2316.871698
Name: 분양가격, dtype: float64

이번에는 분양가격이 100보다 작은 행은 제거해 보겠습니다.

In [90]:
df.loc[df['분양가격'] < 100]

지역명           규모구분    연도   월  분양가격
28    광주   85㎡초과 102㎡이하  2015  10     0
29    광주         102㎡초과  2015  10     0
34    대전         102㎡초과  2015  10     0
81    제주          60㎡이하  2015  10     0
113   광주   85㎡초과 102㎡이하  2015  11     0
...   ..            ...   ...  ..   ...
4461  세종          60㎡이하  2020   2     0
4488  전남   85㎡초과 102㎡이하  2020   2     0
4493  경북   85㎡초과 102㎡이하  2020   2     0
4499  경남         102㎡초과  2020   2     0
4503  제주   85㎡초과 102㎡이하  2020   2     0

[320 rows x 5 columns]

[tip] 특정 조건에 만족하는 행을 제거하고자 할때는
1. index를 list로 가져온다.
2. drop을 활용하여 행을 제거한다.

In [91]:
idx = df.loc[df['분양가격'] < 100].index

In [92]:
df = df.drop(idx) 
df

지역명           규모구분    연도   월  분양가격
0     서울             전체  2015  10  5841
1     서울          60㎡이하  2015  10  5652
2     서울    60㎡초과 85㎡이하  2015  10  5882
3     서울   85㎡초과 102㎡이하  2015  10  5721
4     서울         102㎡초과  2015  10  5879
...   ..            ...   ...  ..   ...
4498  경남   85㎡초과 102㎡이하  2020   2  3247
4500  제주             전체  2020   2  3955
4501  제주          60㎡이하  2020   2  4039
4502  제주    60㎡초과 85㎡이하  2020   2  3962
4504  제주         102㎡초과  2020   2  3601

[4185 rows x 5 columns]

In [93]:
df.count()

지역명     4185
규모구분    4185
연도      4185
월       4185
분양가격    4185
dtype: int64

df의 행의 갯수가 줄어든 것이 보이시나요?
  - df.count() 값이 4185개가 출력이 되어야 합니다!!



다시 한 번 지역명으로 group을 묶어 분양가격을 확인해 보겠습니다.

In [94]:
df.groupby('지역명')['분양가격'].mean()

지역명
강원    2412.642023
경기    4072.667925
경남    2814.376923
경북    2547.486166
광주    3049.028169
대구    3663.335938
대전    3128.433333
부산    3679.920755
서울    7225.762264
세종    2984.004000
울산    3043.503145
인천    3633.275862
전남    2304.969349
전북    2348.648855
제주    3432.795652
충남    2501.604743
충북    2316.871698
Name: 분양가격, dtype: float64

count()로 데이터 갯수도 확인해 볼까요?

In [95]:
df.groupby('지역명')['분양가격'].count()

지역명
강원    257
경기    265
경남    260
경북    253
광주    213
대구    256
대전    210
부산    265
서울    265
세종    250
울산    159
인천    261
전남    261
전북    262
제주    230
충남    253
충북    265
Name: 분양가격, dtype: int64

지역 별 최고로 비싼 분양가는 어떻게 될까요?

In [96]:
df.groupby('지역명')['분양가격'].max()

지역명
강원     3906
경기     5670
경남     4303
경북     3457
광주     4881
대구     5158
대전     4877
부산     4623
서울    13835
세종     3931
울산     3594
인천     5188
전남     3053
전북     3052
제주     5462
충남     3201
충북     2855
Name: 분양가격, dtype: int64

7. 연도별로 분양가격을 확인해 봅시다

In [97]:
df.groupby('연도')['분양가격'].mean()

연도
2015    2788.707819
2016    2934.250000
2017    3143.311795
2018    3326.951034
2019    3693.422149
2020    3853.960526
Name: 분양가격, dtype: float64

8. 비펏 테이블을 활용해서 볼까요?

  - 행 인덱스 : 연도
  - 열 인덱스 : 규모구분
  - 값 : 분양가격

In [98]:
pd.pivot_table(df, index = '연도', columns = '규모구분', values= '분양가격' )

규모구분       102㎡초과        60㎡이하   60㎡초과 85㎡이하   85㎡초과 102㎡이하           전체
연도                                                                      
2015  2980.977778  2712.583333   2694.490196    2884.395833  2694.862745
2016  3148.099476  2848.144279   2816.965686    3067.380435  2816.073529
2017  3427.649746  3112.538071   2981.950980    3204.075145  3008.279412
2018  3468.355932  3286.184783   3227.458128    3467.184211  3235.098522
2019  4039.854839  3486.910112   3538.545918    3933.538462  3515.974490
2020  4187.566667  3615.968750   3594.852941    4532.090909  3603.911765

9. 연도별, 규모별 가격을 알아 볼까요?(multi-index)

In [ ]:
df.groupby(['연도', '규모구분'])['분양가격'].mean()

연도    규모구분         
2015   102㎡초과          2980.977778
       60㎡이하           2712.583333
       60㎡초과 85㎡이하     2694.490196
       85㎡초과 102㎡이하    2884.395833
      전체               2694.862745
2016   102㎡초과          3148.099476
       60㎡이하           2848.144279
       60㎡초과 85㎡이하     2816.965686
       85㎡초과 102㎡이하    3067.380435
      전체               2816.073529
2017   102㎡초과          3427.649746
       60㎡이하           3112.538071
       60㎡초과 85㎡이하     2981.950980
       85㎡초과 102㎡이하    3204.075145
      전체               3008.279412
2018   102㎡초과          3468.355932
       60㎡이하           3286.184783
       60㎡초과 85㎡이하     3227.458128
       85㎡초과 102㎡이하    3467.184211
      전체               3235.098522
2019   102㎡초과          4039.854839
       60㎡이하           3486.910112
       60㎡초과 85㎡이하     3538.545918
       85㎡초과 102㎡이하    3933.538462
      전체               3515.974490
2020   102㎡초과          4187.566667
       60㎡이하           3615.968750
       60㎡초과 85㎡이하     3594.852941


예쁘게 출력이 안되서 보기가 힘들때는 pd.DataFrame()으로 한번 더 감싸주면 됩니다.

In [ ]:
pd.DataFrame(df.groupby(['연도', '규모구분'])['분양가격'].mean())

분양가격
연도   규모구분                      
2015  102㎡초과        2980.977778
      60㎡이하         2712.583333
      60㎡초과 85㎡이하   2694.490196
      85㎡초과 102㎡이하  2884.395833
     전체             2694.862745
2016  102㎡초과        3148.099476
      60㎡이하         2848.144279
      60㎡초과 85㎡이하   2816.965686
      85㎡초과 102㎡이하  3067.380435
     전체             2816.073529
2017  102㎡초과        3427.649746
      60㎡이하         3112.538071
      60㎡초과 85㎡이하   2981.950980
      85㎡초과 102㎡이하  3204.075145
     전체             3008.279412
2018  102㎡초과        3468.355932
      60㎡이하         3286.184783
      60㎡초과 85㎡이하   3227.458128
      85㎡초과 102㎡이하  3467.184211
     전체             3235.098522
2019  102㎡초과        4039.854839
      60㎡이하         3486.910112
      60㎡초과 85㎡이하   3538.545918
      85㎡초과 102㎡이하  3933.538462
     전체             3515.974490
2020  102㎡초과        4187.566667
      60㎡이하         3615.968750
      60㎡초과 85㎡이하   3594.852941
      85㎡초과 102㎡이하  4532.090909
     전체             3603.911765